helix-term/src/commands.rs

```diff
@@ -2097,54 +2097,97 @@ fn search_impl(
     // it out, we need to add it back to the position of the selection.
     let doc = doc!(editor).text().slice(..);
 
-    // use find_at to find the next match after the cursor, loop around the end
+    let mut all_matches = regex.find_iter(doc.regex_input()).enumerate().peekable();
+
+    if all_matches.peek().is_none() {
+        if show_warnings {
+            editor.set_error("No matches");
+        }
+        return;
+    }
+
+    // We will get the number of the current match and total matches from
+    // `all_matches`. So, here we look in the iterator until we find exactly
+    // the match we were looking for (either after or behind `start`). In the
+    // `Backward` case, in particular, we need to look the match ahead to know
+    // if this is the one we need.
+
     // Careful, `Regex` uses `bytes` as offsets, not character indices!
     let mut mat = match direction {
-        Direction::Forward => regex.find(doc.regex_input_at_bytes(start..)),
-        Direction::Backward => regex.find_iter(doc.regex_input_at_bytes(..start)).last(),
+        Direction::Forward => all_matches.by_ref().find(|&(_, m)| m.start() >= start),
+        Direction::Backward => {
+            let one_behind = std::iter::once(None).chain(all_matches.by_ref().map(Some));
+            one_behind
+                .zip(regex.find_iter(doc.regex_input()))
+                .find(|&(_, m1)| m1.start() >= start)
+                .map(|(m0, _)| m0)
+                .unwrap_or(None)
+        }
     };
 
-    if mat.is_none() {
-        if wrap_around {
-            mat = match direction {
-                Direction::Forward => regex.find(doc.regex_input()),
-                Direction::Backward => regex.find_iter(doc.regex_input_at_bytes(start..)).last(),
-            };
+    if mat.is_none() && !wrap_around {
+        if show_warnings {
+            editor.set_error("No more matches");
         }
+        return;
+    }
+
+    // If we didn't find a match before, lets wrap the search.
+    if mat.is_none() {
         if show_warnings {
-            if wrap_around && mat.is_some() {
-                editor.set_status("Wrapped around document");
-            } else {
-                editor.set_error("No more matches");
-            }
+            editor.set_status("Wrapped around document");
         }
+
+        let doc = doc!(editor).text().slice(..);
+        all_matches = regex.find_iter(doc.regex_input()).enumerate().peekable();
+        mat = match direction {
+            Direction::Forward => all_matches.by_ref().next(),
+            Direction::Backward => all_matches.by_ref().last(),
+        };
     }
 
+    let (idx, mat) = mat.unwrap();
+    let last_idx = match all_matches.last() {
+        None => idx,
+        Some((last_idx, _)) => last_idx,
+    };
```

Comment:
This is inefficient and also the reason we didn't implement it in the first place: it requires scanning through the entire document just to find a total count, and this is a significant hit on very large files/lots of matches. (Because of this vim only says 99+ if there's more than 100 matches).

Reply:
That's fair. I thought about this myself and I happened to try in a big file (hundreds of megabytes) and it was pretty responsive. That made sense because we have to go through the whole file in case we wrap the backward search and that seemed to be working fine. But I guess I don't know exactly what use case you have in mind, maybe a gigabytes long file with a slow machine, which is valid.

If we really want to cap the number of matches, I was thinking about two solutions for this:
1. We can go the `vim` route and cap it to 100 matches.
2. We add a configuration to let people decide the number of matches they are willing to tolerate.

Let me know your preference. Either way is fine.

Reply:
modern computers are absolutely fast enough, and the 100 matches restriction is a stretch. could be the default if it's that important, but *has to* be configurable to be more. we shouldn't repeat vim's history

Reply:
Actually I tried this and it's really responsive. I think the option should be available at least. 

This is a freaking great feature! TY @useche regardless if it gets merged or not.

Reply:
yeah worst case scenario I'll just remove the cap in my personal fork hehehe so I'll be happy to see this merged regardless, very very needed feature imo

Reply:
> This is inefficient and also the reason we didn't implement it in the first place: it requires scanning through the entire document just to find a total count, and this is a significant hit on very large files/lots of matches. (Because of this vim only says 99+ if there's more than 100 matches).

I added an option to cap the maximum number of matches counted that by default is 100. The option can be set to `"none"` to count all the matches. Please take a look.